convert.py

```diff
@@ -201,6 +202,30 @@ def load(model_plus: 'ModelPlus') -> 'Params':
         return params
 
 
+class XgenVocab:
+    def __init__(self, path: Path) -> None:
+        self.fname_tokenizer = path
+        self.fname_added_tokens = None
+        path = str((path / "tokenization_xgen.py").absolute())
+        spec = importlib.util.spec_from_file_location(path, path)
+        module = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(module)
+        self.xt = module.XgenTokenizer()
+        self.vocab_size_base: int = self.xt.vocab_size
+        self.vocab_size: int = self.xt.vocab_size
+        self.added_tokens_list = []
+
+    def all_tokens(self) -> Iterable[Tuple[bytes, float]]:
+        for index in range(0, self.vocab_size_base):
+            token = self.xt._convert_id_to_token(index)
+            yield (token, float(index))
+        for index in range(self.vocab_size_base, self.vocab_size):
+            yield (b'', float(index))
```

Comment:
```suggestion
            yield (b'<|unk|>', float(index))
```

🤷

Reply:
I've tried the latest commit and this is what I got.

```
$ python3 convert.py models/xgen-4k-7b-orig --outtype f16
Loading model file models/xgen-4k-7b-orig/pytorch_model-00001-of-00003.bin
Loading model file models/xgen-4k-7b-orig/pytorch_model-00001-of-00003.bin
Loading model file models/xgen-4k-7b-orig/pytorch_model-00002-of-00003.bin
Loading model file models/xgen-4k-7b-orig/pytorch_model-00003-of-00003.bin
params: n_vocab:51200 n_embd:4096 n_mult:256 n_head:32 n_layer:32
Writing vocab...
Traceback (most recent call last):
  File "/home/user/llama.cpp/convert.py", line 1255, in <module>
    main()
  File "/home/user/llama.cpp/convert.py", line 1250, in main
    OutputFile.write_all(outfile, params, output_type, model, vocab)
  File "/home/user/llama.cpp/convert.py", line 1041, in write_all
    of.write_vocab(vocab)
  File "/home/user/llama.cpp/convert.py", line 1022, in write_vocab
    self.fout.write(text)
TypeError: a bytes-like object is required, not 'str'
```

Reply:
I can confirm convert.py and quantize both work, thanks @smdesai !!

Btw, you still need to change EOS, BOS and NL token ID in llama.cpp to make it inference correctly.

Reply:
Can you share what you changed in llama.cpp?

Reply:
@tmm1 Here is what I modified in llama.cpp:

```c
llama_token llama_token_bos() {
    return 50256;
}

llama_token llama_token_eos() {
    return 50256;
}

llama_token llama_token_nl() {
    return 198;
}
```